plugins/gssapi.c

```diff
@@ -1114,32 +1114,63 @@ gssapi_server_mech_authneg(context_t *text,
     gss_OID mech_type;
     gss_channel_bindings_t bindings = GSS_C_NO_CHANNEL_BINDINGS;
     struct gss_channel_bindings_struct cb = {0};
+    unsigned int rl; /* Required storage by getopt() */
+    const char *service_principal_opt = NULL; /* The configured service principal name */
+    const char *any_principal = NULL; /* Do we accept any principal in the server keytab? */
+    unsigned char accept_any = 0;
+    const sasl_utils_t *utils = text->utils;
+#ifdef GSS_USE_LOAD_FROM
+    char* keytab = NULL; /* The configured credentials cache file path */
+
+    utils->getopt(utils->getopt_context, "GSSAPI", "keytab", (const char**)&keytab, &rl);
+#endif
 
     input_token = &real_input_token;
     output_token = &real_output_token;
     output_token->value = NULL; output_token->length = 0;
     input_token->value = NULL; input_token->length = 0;
-    
-    if (text->server_name == GSS_C_NO_NAME) { /* only once */
-	if (params->serverFQDN == NULL
-	    || strlen(params->serverFQDN) == 0) {
-	    SETERROR(text->utils, "GSSAPI Failure: no serverFQDN");
-	    sasl_gss_free_context_contents(text);
-	    return SASL_FAIL;
+
+    utils->getopt(utils->getopt_context,
+		  "GSSAPI", "accept_any_principal", (const char**) &any_principal, &rl);
+    if (any_principal &&
+        ((rl >=1 && (any_principal[0] == '1' ||
+		     any_principal[0] == 'y' ||
+		     any_principal[0] == 't')) ||
+	 (rl >=2 && (any_principal[0] == 'o' && any_principal[1] == 'n'))))
+	accept_any = 1;
+
+    if (!accept_any && text->server_name == GSS_C_NO_NAME) { /* only once */
+	utils->getopt(utils->getopt_context,
+                      "GSSAPI", "service_principal", (const char**) &service_principal_opt, &rl);
+	if (service_principal_opt) {
+            name_token.length = strlen(service_principal_opt);
+	} else {
+	    if (params->serverFQDN == NULL
+		|| strlen(params->serverFQDN) == 0) {
+		SETERROR(text->utils, "GSSAPI Failure: no serverFQDN");
+		sasl_gss_free_context_contents(text);
+		return SASL_FAIL;
+	    }
+	    name_token.length = strlen(params->service) + 1 + strlen(params->serverFQDN);
 	}
-	name_token.length = strlen(params->service) + 1 + strlen(params->serverFQDN);
+
 	name_token.value = (char *)params->utils->malloc((name_token.length + 1) * sizeof(char));
 	if (name_token.value == NULL) {
 	    MEMERROR(text->utils);
 	    sasl_gss_free_context_contents(text);
 	    return SASL_NOMEM;
 	}
-	sprintf(name_token.value,"%s@%s", params->service, params->serverFQDN);
+
+	if (service_principal_opt) {
+	    strncpy(name_token.value, service_principal_opt, name_token.length + 1);
+	} else {
+	    sprintf(name_token.value,"%s@%s", params->service, params->serverFQDN);
+	}
 
 	GSS_LOCK_MUTEX_CTX(params->utils, text);
 	maj_stat = gss_import_name (&min_stat,
 				    &name_token,
-				    GSS_C_NT_HOSTBASED_SERVICE,
+				    service_principal_opt ? GSS_KRB5_NT_PRINCIPAL_NAME : GSS_C_NT_HOSTBASED_SERVICE,
```

Comment:
Why mixing types here?
The service principal name should be a GSSAPI name specified as svc@name.

If you want to support *kerberos* principal names, you should name the parameter differently, but I hardly see a reason to do that.

Reply:
GSS_C_NT_HOSTBASED_SERVICE can't express all valid KRB5 principal names, but I don't care enough to keep it in. Switched to always using GSS_C_NT_HOSTBASED_SERVICE.